src/include/common/execution_thread_pool.h

```diff
@@ -0,0 +1,270 @@
+    }
+  }
+
+  /**
+   * Destructor. Wake up all workers and let them finish before it's destroyed.
+   */
+  ~ExecutionThreadPool() override {
+    shutting_down_ = true;
+    for (std::vector<TerrierThread *> vector : workers_) {  // NOLINT
+      for (TerrierThread *t : vector) {
+        bool result UNUSED_ATTRIBUTE = thread_registry_.operator->()->StopTask(
+            this, common::ManagedPointer(static_cast<DedicatedThreadTask *>(t)));
+        TERRIER_ASSERT(result, "StopTask should succeed");
+      }
+    }
+  }
+
+  /**
+   * SubmitTask allows for a user to submit a task to the given NUMA region
+   * @param promise a void promise pointer that will be set when the task has been executed
+   * @param task a void to void function that is the task to be executed
+   * @param numa_hint a hint as to which NUMA region would be ideal for this task to be executed on, default is any
+   */
+  void SubmitTask(std::promise<void> *promise, const std::function<void()> &task,
+                  storage::numa_region_t numa_hint = storage::UNSUPPORTED_NUMA_REGION) {
+    if (numa_hint == storage::UNSUPPORTED_NUMA_REGION) {
+      numa_hint = static_cast<storage::numa_region_t>(0);
+    }
+    Task t({promise, task});
+    task_queue_[static_cast<int16_t>(numa_hint)].push(t);
+    task_cv_.notify_all();
+  }
+
+  /**
+   * Get the number of worker threads in this pool
+   *
+   * @return The number of worker threads
+   */
+  uint32_t NumWorkers() const { return total_workers_; }
+
+ private:
+  // Private thread co-class
+  class TerrierThread : public DedicatedThreadTask {
+   public:
+    TerrierThread(int cpu_id, ExecutionThreadPool *pool) : pool_(pool), cpu_id_(cpu_id) {
+#ifndef __APPLE__
+      cpu_set_t mask;
+      CPU_ZERO(&mask);
+      CPU_SET(cpu_id_, &mask);
+      int result UNUSED_ATTRIBUTE = sched_setaffinity(0, sizeof(cpu_set_t), &mask);
+      TERRIER_ASSERT(result == 0, "sched_setaffinity should succeed");
+      numa_region_ = storage::UNSUPPORTED_NUMA_REGION;
+      if (numa_available() >= 0) {
+        numa_region_ = static_cast<storage::numa_region_t>(numa_node_of_cpu(cpu_id));
+      }
+      if (static_cast<int16_t>(numa_region_) < 0) {
+        numa_region_ = static_cast<storage::numa_region_t>(0);
+      }
+#else
+      // TODO(emmanuee) figure out processor_assign and put here
+      numa_region_ = static_cast<storage::numa_region_t>(0);
+#endif
+      pool_->total_workers_++;
+    }
+    ~TerrierThread() override = default;
+
+    void RunNextTask() {
+      while (true) {
+        for (int16_t i = 0; i < pool_->num_regions_; i++) {
+          auto index = (static_cast<int16_t>(numa_region_) + i) % pool_->num_regions_;
+          Task task;
+          if (!pool_->task_queue_[index].try_pop(task)) {
+            continue;
+          }
+
+          status_ = ThreadStatus::BUSY;
+          task.second();
+          task.first->set_value();
+          status_ = ThreadStatus::SWITCHING;
+          return;
+        }
+
```

Comment:
Some comments here would be helpful - if I understand it correctly, it first tries to run a task, and wait for tasks if there's no task present.

Reply:
It has changed a fair amount since we added coroutines but it is now (hopefully) easily understandable now